Code/GraphMol/SmilesParse/smatest.cpp

```diff
@@ -2671,11 +2671,81 @@ void testGithub2142() {
 
   BOOST_LOG(rdInfoLog) << "done" << std::endl;
 }
+
+void testGithub2565() {
+  BOOST_LOG(rdInfoLog) << "-------------------------------------" << std::endl;
+  BOOST_LOG(rdInfoLog)
+      << "Testing Github #2565: Chirality reversed on SMARTS generation."
+      << std::endl;
+
+  std::vector<std::string> smiles({
+      R"(O=C1C[C@]([H])1F)",
+      R"(Cl[C@@H]1CCC1=O)",
+      R"(N1CCN=C1[C@H]1CCCc2ccccc21)",
+      R"(C[C@@H](Cl)[C@H]1CC[C@@H](Cl)CC1)",
+      R"(Fc1cn([C@@H]2CCCO2)c(=O)[nH]c1=O)",
+      R"([C@@]1(C)(C(C)(C)C)O[C@@H](CN)[C@H](C[NH3+])O1)",
+
+      // these are Ok
+      R"(O=C1C[C@](Cl)1F)",
+      R"(Br[C@@H](Cl)F)"});
+
+  for (const auto &smi : smiles) {
+    auto *mol = SmilesToMol(smi);
+    const std::string smarts = MolToSmarts(*mol, true);
+    auto *query = SmartsToMol(smarts);
+
+    std::vector<MatchVectType> matches;
+    TEST_ASSERT(SubstructMatch(*mol, *query, matches, true, true, true, false));
+  }
+
+  BOOST_LOG(rdInfoLog) << "done" << std::endl;
+}
+
+void testGithub2568() {
+  BOOST_LOG(rdInfoLog) << "-------------------------------------" << std::endl;
+  BOOST_LOG(rdInfoLog) << "Testing Github #2568: SMARTS parser reverses "
+                          "chirality on leading chiral atoms. "
+                       << std::endl;
+
+  std::vector<std::string> smiles({
+      R"(O=C1C[C@]([H])1F)",
+      R"([C@H](O)(F)Cl)",
+
+      /*
+       * This is a tough one: it won't keep chirality in SMILES.
+       * Let's skip it for now.
+       * R"([C@](O)(F)Cl)",
+       */
+
+      R"([H][C@](O)(F)Cl)",
+      R"([C@]([H])(O)(F)Cl)",
+      R"([C@](O)([H])(F)Cl)",
+      R"([C@](O)(F)([H])Cl)",
+      R"([C@](O)(F)(Cl)[H])",
+
+      // these are Ok
+      R"(O=C1C[C@](Cl)1F)",
+      R"(Br[C@@H](Cl)F)",
+      R"(Fc1cn([C@@H]2CCCO2)c(=O)[nH]c1=O)",
+      R"([C@@]1(C)(C(C)(C)C)O[C@@H](CN)[C@H](C[NH3+])O1)"});
+
+  for (const auto &smi : smiles) {
+    auto *mol = SmilesToMol(smi);
+    auto *query = SmartsToMol(smi);
+
+    MolOps::addHs(*mol);  // We have explicit H in the queries!
+
+    std::vector<MatchVectType> matches;
+    TEST_ASSERT(SubstructMatch(*mol, *query, matches, true, true, true, false));
```

Comment:
One of the habits I picked up from @bp-kelley is having local variables with the names of the optional arguments so that long streams of bools like we have here are actually understandable:
```suggestion
    bool uniquify=true,recursionPossible=true, useChirality=true;
    TEST_ASSERT(SubstructMatch(*mol, *query, matches, uniquify, recursionPossible, useChirality));
```

Reply:
I intentionally left off the last argument (`useQueryQueryMatches`) since it was left in its default value.

Reply:
That makes sense. I'll update with it, although I don't like declaring more than one variable in the same line.

Reply:
me neither, I was just too lazy to do the suggestion across multiple lines in the github UI.

Reply:
heh, heh, no problem.